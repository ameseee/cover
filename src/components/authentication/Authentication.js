import React, { Component } from 'react';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import firebase from '../../firebase';

class Authentication extends Component {
  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      newUser: false,
    };
  }

  createAccount = (newUser) => {
    Object.entries(newUser).map(([key, value]) => {
      Object.assign({id: key}, value);
    });
  }

  componentDidMount() {
    const createRef = firebase.database().ref('users');

    createRef.on('value', (snapshot) => {
      const newUser = snapshot.val();

      if (newUser) this.createAccount(newUser);
    });
  }

  handleChange(key, event) {
    this.setState({ [key]: event.target.value });
  }

  createAccountClick = (email, password) => {
    this.clearState();
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(login => {
        this.props.setCurrentUser(login.uid);
        this.props.signIn(true);
        this.props.history.push('/main');
      })
      .catch( error => {
        throw new Error(error);
      });
  };

  signInClick = (email, password) => {
    this.clearState();
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(login => {
        this.props.setCurrentUser(login.uid);
        this.props.signIn(true);
        this.props.history.push('/main');
      })
      .catch(error => {
        this.props.history.push('/auth');
        alert(error);
      });
  };

  clearState() {
    this.setState({
      username: '',
      password: '',
    });
  }

  switchToOtherForm = () => {
    this.setState({ newUser: !this.state.newUser });
  }

  render() {
    return (
      <div className="authentication-section">
        <section
          className={classnames(
            'create-account-section', this.state.newUser ? '' : 'hidden')}
        >
          <h3 className="create-account-title">Create Your Account</h3>
          <div>
            <input
              className="name"
              type="text"
              value={this.state.username}
              onChange={this.handleChange.bind(this, 'username')}
              placeholder="Name/UserName"
            />
            <input
              className="password"
              type="text"
              value={this.state.password}
              onChange={this.handleChange.bind(this, 'password')}
              placeholder="Password"
            />
            <button
              className="btn"
              onClick={() => this.createAccountClick(
                this.state.username,
                this.state.password)}
            >
              Create Account
            </button>
          </div>
        </section>
        <section
          className={classnames(
            'sign-in-section',
            this.state.newUser ? 'hidden' : '')}
        >
          <h3 className="sign-in-title">Sign In</h3>
          <div>
            <input
              className="name"
              type="text"
              value={this.state.username}
              onChange={this.handleChange.bind(this, 'username')}
              placeholder="Name/UserName"
            />
            <input
              className="password"
              type="text"
              value={this.state.password}
              onChange={this.handleChange.bind(this, 'password')}
              placeholder="Password"
            />
            <button
              className="btn"
              onClick={() => this.signInClick(
                this.state.username,
                this.state.password)}>
              <Link
                to='/'
                className="link-to-main"
              >
                Sign In
              </Link>
            </button>
          </div>
        </section>
        <article className="other-card-section">
          <h4
            className="already">{
              this.state.newUser
                ? 'Already have an account?'
                : 'Don\'t have an accout yet?'}
          </h4>
          <button
            className="other-card-btn" onClick={this.switchToOtherForm}
          >
            {this.state.newUser ? 'Sign In' : 'Sign Up'}
          </button>
        </article>
      </div>
    );
  }
}

Authentication.propTypes = {
  currentUser: PropTypes.string,
  setCurrentUser: PropTypes.func,
  signIn: PropTypes.func,
  history: PropTypes.object,
};

export default Authentication;
