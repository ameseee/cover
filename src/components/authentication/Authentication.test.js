import React from 'react';
import Authentication from './Authentication';
import { shallow } from 'enzyme';

describe('<Authentication />', () => {
  let wrapper;
  let mockFn = jest.fn();

  beforeEach( () => {
    const wrapper = shallow(<Authentication />);
  });

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it.skip('should have a empty default state of username', () => {
    expect(wrapper.state().username).toEqual('');
  });

  it.skip('should have a empty default state of password', () => {
    expect(wrapper.state().password).toEqual('');
  });

  it.skip('should have a false default state of newUser', () => {
    expect(wrapper.state().newUser).toEqual(false);
  });

  it.skip('should render a title', () => {
    const title = wrapper.find('h3');

    expect(title.length).toEqual(1);
  });

  it.skip('should render two inputs', () => {
    const input = wrapper.find('input');

    expect(input.length).toEqual(2);
  });

  it.skip('should change state on input', () => {
    const username = wrapper.find('.name');
    const password = wrapper.find('.password');

    username.simulate('change', { target: { value: 'amy@gmail.com' } });
    password.simulate('change', { target: { value: 'password' } });
    expect(wrapper.state().username).toEqual('amy@gmail.com');
    expect(wrapper.state().password).toEqual('password');
  });

  it.skip('should render a button to sign up/in', () => {
    const button = wrapper.find('btn');

    expect(button.length).toEqual(1);
  });

  it.skip('should run a function to sign up/in', () => {
    const button = wrapper.find('btn');

    button.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it.skip('should clear input and state once user signs in/up', () => {
    const button = wrapper.find('btn');

    button.simulate('click');
    expect(wrapper.state().username).toEqual('');
    expect(wrapper.state().password).toEqual('');
  });

  it.skip('should render a prompt to toggle forms', () => {
    const prompt = wrapper.find('h4');

    expect(prompt.length).toEqual(1);
  });

  it.skip('should render a button to toggle sign in/up', () => {
    const button = wrapper.find('other-card-btn');

    expect(button.length).toEqual(1);
  });

  it.skip('should run a function to switch forms', () => {
    const button = wrapper.find('other-card-btn');

    button.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it.skip('should change state when other form button is clicked', () => {
    const button = wrapper.find('other-card-btn');

    expect(wrapper.state().newUser).toEqual(false);
    button.simulate('click');
    expect(wrapper.state().newUser).toEqual(true);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
});
