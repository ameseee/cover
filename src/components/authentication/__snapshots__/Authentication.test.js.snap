// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authentication /> should match snapshot 1`] = `undefined`;
